From b4f15cb2c75f87504e2246a0205d483ed5fcc5f7 Mon Sep 17 00:00:00 2001
From: Naveen S <quic_naves@quicinc.com>
Date: Thu, 5 Sep 2024 11:19:07 +0530
Subject: [PATCH] sigma_dut: Restructuring to extend multi link support

Currently only two links are being supported via sigma
and two variables are being used for each link for every feature
which limits in reusing the varibles at certain times and limits
to extend the multi link support if needed.

Providing support by adding a new structure "mld_ap" which
will hold all the link specific variables which eases the code
accessibililty and reusability.

Signed-off-by: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
Signed-off-by: Naveen S <quic_naves@quicinc.com>
---
 ap.c        | 185 ++++++++++++++++++++--------------------------------
 sigma_dut.h | 124 +++++++++++++++++++++--------------
 2 files changed, 146 insertions(+), 163 deletions(-)

--- a/ap.c
+++ b/ap.c
@@ -984,6 +984,18 @@ static int wcn_set_txbf_periodic_ndp(str
 }
 #endif /* NL80211_SUPPORT */
 
+/* To obtain the dut's mld link_id based on the band information */
+int get_link_id_by_band(struct sigma_dut *dut, enum if_band band)
+{
+	int link_id;
+
+	for (link_id = 0; link_id < NUM_OF_LINKS; link_id++) {
+		if (check_band(dut, band, link_id))
+			break;
+	}
+
+	return link_id;
+}
 
 static enum sigma_cmd_result cmd_ap_set_wireless(struct sigma_dut *dut,
 						 struct sigma_conn *conn,
@@ -992,15 +1004,14 @@ static enum sigma_cmd_result cmd_ap_set_
 	/* const char *name = get_param(cmd, "NAME"); */
 	/* const char *ifname = get_param(cmd, "INTERFACE"); */
 	const char *val;
-	unsigned int wlan_tag = 1;
-	int chwidth_check = 0;
+	unsigned int wlan_tag = 0;
 	const char *ifname = get_main_ifname(dut);
 	char buf[128];
 	int is_6g = 0;
 	enum ap_chwidth chwidth;
 	int subeamformermode = 0;
-	int dtim_check = 0;
 	int iface_check = 0;
+	int link_id = dut->num_links - 1;
 
 	/* Allow program to be overridden if specified in the ap_set_wireless
 	 * to support some 60 GHz test scripts where the program may be 60 GHz
@@ -1059,7 +1070,6 @@ static enum sigma_cmd_result cmd_ap_set_
 
 			val = get_param(cmd, "Channel");
 			if(val) {
-				dut->ap_channel = atoi(val);
 				dut->mbss_set[mbssid_set_idx].conf[mbss_conf_idx].channel = atoi(val);
 			}
 
@@ -1092,13 +1102,13 @@ static enum sigma_cmd_result cmd_ap_set_
 			if (val) {
 				if (strcasecmp(val, "5G") == 0) {
 					dut->ap_interface_5g = 1;
-					dut->ap_band = BAND_5G;
+					dut->mbss_set[mbssid_set_idx].conf[mbss_conf_idx].ap_band = BAND_5G;
 				} else if (strcasecmp(val, "6G") == 0) {
 					dut->ap_interface_6g = 1;
-					dut->ap_band = BAND_6G;
+					dut->mbss_set[mbssid_set_idx].conf[mbss_conf_idx].ap_band = BAND_6G;
 				} else if (strcasecmp(val, "24G") == 0) {
 					dut->ap_interface_2g = 1;
-					dut->ap_band = BAND_2G;
+					dut->mbss_set[mbssid_set_idx].conf[mbss_conf_idx].ap_band = BAND_2G;
 				}
 			}
 
@@ -1200,92 +1210,68 @@ static enum sigma_cmd_result cmd_ap_set_
 		}
 
 		val = get_param(cmd, "CHANNEL");
+		if (val)
+			dut->mld[link_id].ap_channel = atoi(val);
+
+		val = get_param(cmd, "WIDTH");
 		if (val) {
-			if (dut->ap_is_mld) {
-				dut->ap_tag_channel[0] = atoi(val);
-				dut->ap_channel_1 = atoi(val);
-			} else if (wlan_tag == 1) {
-				const char *pos;
-
-				dut->ap_channel = atoi(val);
-				pos = strchr(val, ';');
-				if (pos) {
-					pos++;
-					dut->ap_channel_1 = atoi(pos);
-				}
-			} else {
-				dut->ap_tag_channel[wlan_tag - 2] = atoi(val);
-				dut->ap_channel_1 = atoi(val);
+			if (strcasecmp(val, "20") == 0)
+				chwidth = AP_20;
+			else if (strcasecmp(val, "40") == 0)
+				chwidth = AP_40;
+			else if (strcasecmp(val, "80") == 0)
+				chwidth = AP_80;
+			else if (strcasecmp(val, "160") == 0)
+				chwidth = AP_160;
+			else if (strcasecmp(val, "320") == 0)
+				chwidth = AP_320;
+			else if (strcasecmp(val, "80plus80") == 0) {
+				dut->ap_80plus80 = 1;
+				chwidth = AP_80_80;
+			} else if (strcasecmp(val, "Auto") == 0)
+				chwidth = AP_AUTO;
+			else {
+				send_resp(dut, conn, SIGMA_INVALID,
+						"errorCode,Unsupported WIDTH");
+				return STATUS_SENT;
 			}
+			dut->mld[link_id].ap_chwidth = chwidth;
 		}
 
-		if (!chwidth_check) {
-			val = get_param(cmd, "WIDTH");
-			if (val) {
-				if (strcasecmp(val, "20") == 0)
-					chwidth = AP_20;
-				else if (strcasecmp(val, "40") == 0)
-					chwidth = AP_40;
-				else if (strcasecmp(val, "80") == 0)
-					chwidth = AP_80;
-				else if (strcasecmp(val, "160") == 0)
-					chwidth = AP_160;
-				else if (strcasecmp(val, "320") == 0)
-					chwidth = AP_320;
-				else if (strcasecmp(val, "80plus80") == 0) {
-					dut->ap_80plus80 = 1;
-					chwidth = AP_80_80;
-				} else if (strcasecmp(val, "Auto") == 0)
-					chwidth = AP_AUTO;
-				else {
-					send_resp(dut, conn, SIGMA_INVALID,
-							"errorCode,Unsupported WIDTH");
-					return STATUS_SENT;
-				}
-				/*
-				 * TODO: Need to understand whether similar
-				 * logic to save chwidth based on link_id here
-				 * as well
-				 */
-				if (wlan_tag > 1)
-					dut->ap_chwidth_1 = chwidth;
-				else
-					dut->ap_chwidth = chwidth;
+		val = get_param(cmd, "ChnlFreq");
+		if (val) {
+			if (atoi(val) >= 5935 && atoi(val) <= 7115) {
+				is_6g = 1;
+				dut->cfreq = atoi(val);
+				dut->ap_band_6g = 1;
+				dut->ap_mbo = 1;
+			} else {
+				dut->ap_band_6g = 0;
 			}
 		}
 
-		if ((wlan_tag > 1 && dut->ap_band_1 == BAND_UNKNOWN) ||
-			(dut->ap_is_mld && dut->ap_band_1 == BAND_UNKNOWN)) {
-			if (is_6g)
-				dut->ap_band_1 = BAND_6G;
-			else if (dut->ap_channel_1 >= 1 && dut->ap_channel_1 <= 14)
-				dut->ap_band_1 = BAND_2G;
-			else if (dut->ap_channel_1 >= 36 && dut->ap_channel_1 <= 171)
-				dut->ap_band_1 = BAND_5G;
-
-			if (dut->ap_band_1 != BAND_2G && (dut->ap_mode_1 == AP_11ac ||
-				dut->ap_mode_1 == AP_11ax || dut->ap_mode == AP_11be) &&
-				!chwidth_check)
-				dut->ap_chwidth_1 = AP_80;
-		} else if (dut->ap_band == BAND_UNKNOWN) {
-			if (is_6g)
-				dut->ap_band = BAND_6G;
-			else if (dut->ap_channel >= 1 && dut->ap_channel <= 14)
-				dut->ap_band = BAND_2G;
-			else if (dut->ap_channel >= 36 && dut->ap_channel <= 171)
-				dut->ap_band = BAND_5G;
-
-			if (dut->ap_band != BAND_2G && (dut->ap_mode == AP_11ac ||
-				dut->ap_mode == AP_11ax || dut->ap_mode == AP_11be) &&
-				dut->ap_80plus80 != 1)
-				dut->ap_chwidth = AP_80;
+		if (is_6g) {
+			dut->mld[link_id].ap_band = BAND_6G;
+			dut->ap_interface_6g = 1;
+		} else if (dut->mld[link_id].ap_channel >= 1 &&
+		         dut->mld[link_id].ap_channel <= 14 &&
+			 dut->mld[link_id].ap_band == BAND_UNKNOWN) {
+			dut->mld[link_id].ap_band = BAND_2G;
+			dut->ap_interface_2g = 1;
+		} else if (dut->mld[link_id].ap_channel >= 36 &&
+		         dut->mld[link_id].ap_channel <= 171 &&
+			 dut->mld[link_id].ap_band == BAND_UNKNOWN) {
+			dut->mld[link_id].ap_band = BAND_5G;
+			dut->ap_interface_5g = 1;
 		}
+
 		return SUCCESS_SEND_STATUS;
 	}
 
 	val = get_param(cmd, "WLAN_TAG");
 	if (val) {
 		wlan_tag = atoi(val);
+
 		if (wlan_tag < 1 || wlan_tag > 3) {
 			/*
 			 * The only valid WLAN Tags as of now as per the latest
@@ -1294,9 +1280,13 @@ static enum sigma_cmd_result cmd_ap_set_
 			send_resp(dut, conn, SIGMA_INVALID,
 				  "errorCode,Invalid WLAN_TAG");
 			return STATUS_SENT;
-		}
-		if (wlan_tag > 1)
-			dut->ap_is_dual = 1;
+	        }
+
+	        if (wlan_tag > 1)
+	                dut->ap_is_dual = 1;
+
+	        if (dut->num_links < wlan_tag)
+	                dut->num_links = wlan_tag;
 	}
 
 	val = get_param(cmd, "Interface");
@@ -1314,11 +1304,9 @@ static enum sigma_cmd_result cmd_ap_set_
 		}
 	}
 
-	if (!dtim_check) {
-		val = get_param(cmd, "DTIM");
+	val = get_param(cmd, "DTIM");
 	if (val)
-		dut->dtim = atoi(val);
-	}
+		dut->mld[link_id].dtim = atoi(val);
 
 	val = get_param(cmd, "ChnlFreq");
 	if (val) {
@@ -1343,12 +1331,15 @@ static enum sigma_cmd_result cmd_ap_set_
 		if (!dut->dual_ap_ifname) {
 			dut->ap_is_mld = 1;
 			dut->num_links++;
-
-			if (dut->ap_mode)
-				dut->ap_mode_1 = dut->ap_mode;
 		}
 	}
 
+	link_id = (wlan_tag) ? wlan_tag - 1 : dut->num_links - 1;
+
+	if (dut->ap_is_mld)
+		dut->mld[link_id].ap_mode = dut->mld[dut->num_links - 2].ap_mode;
+
+
 	val = get_param(cmd, "CountryCode");
 	if (val) {
 		if (strlen(val) > sizeof(dut->ap_countrycode) - 1)
@@ -1379,7 +1370,7 @@ static enum sigma_cmd_result cmd_ap_set_
 		if (strlen(val) > sizeof(dut->ap_ssid) - 1)
 			return INVALID_SEND_STATUS;
 
-		if (wlan_tag == 1) {
+		if (wlan_tag <= 1) {
 			/*
 			 * If tag is not specified, it is deemed to be 1.
 			 * Hence tag of 1 is a special case and the values
@@ -1399,24 +1390,8 @@ static enum sigma_cmd_result cmd_ap_set_
 	}
 
 	val = get_param(cmd, "CHANNEL");
-	if (val) {
-		if (dut->ap_is_mld) {
-			dut->ap_tag_channel[0] = atoi(val);
-			dut->ap_channel_1 = atoi(val);
-		} else if (wlan_tag == 1) {
-			const char *pos;
-
-			dut->ap_channel = atoi(val);
-			pos = strchr(val, ';');
-			if (pos) {
-				pos++;
-				dut->ap_channel_1 = atoi(pos);
-			}
-		} else {
-			dut->ap_tag_channel[wlan_tag - 2] = atoi(val);
-			dut->ap_channel_1 = atoi(val);
-		}
-	}
+	if (val)
+		dut->mld[link_id].ap_channel = atoi(val);
 
 	val = get_param(cmd, "ChnlFreq");
 	if (val) {
@@ -1441,8 +1416,8 @@ static enum sigma_cmd_result cmd_ap_set_
 	/* Overwrite the AP channel with DFS channel if configured */
 	val = get_param(cmd, "dfs_chan");
 	if (val) {
-		if (wlan_tag == 1)
-			dut->ap_channel = atoi(val);
+		if (wlan_tag <= 1)
+			dut->mld[link_id].ap_channel = atoi(val);
 		else
 			dut->ap_tag_channel[wlan_tag - 2] = atoi(val);
 	}
@@ -1479,49 +1454,32 @@ static enum sigma_cmd_result cmd_ap_set_
 			return STATUS_SENT;
 		}
 
-		if (wlan_tag > 1) {
-			dut->ap_mode_1 = mode;
-		} else {
-			dut->ap_mode = mode;
-			if ((dut->ap_mode == AP_11ac ||
-			     (dut->ap_mode == AP_11ax && dut->use_5g) ||
-			     (dut->ap_mode == AP_11be && dut->use_5g))
-			     && dut->ap_80plus80 != 1)
-				dut->ap_chwidth = AP_80;
-
-			if (pos) {
-				dut->ap_mode_1 = get_mode(pos);
-				if (dut->ap_mode_1 == AP_inval) {
-					send_resp(dut, conn, SIGMA_INVALID,
-						  "errorCode,Unsupported MODE");
-					free(str);
-					return STATUS_SENT;
-				}
-				if (dut->ap_mode_1 == AP_11ac)
-					dut->ap_chwidth_1 = AP_80;
-				dut->ap_is_dual = 1;
-			}
-		}
+		dut->mld[link_id].ap_mode = mode;
+		if ((dut->mld[link_id].ap_mode == AP_11ac ||
+		     (dut->mld[link_id].ap_mode == AP_11ax && dut->use_5g) ||
+		     (dut->mld[link_id].ap_mode == AP_11be && dut->use_5g))
+		    && dut->ap_80plus80 != 1 && dut->mld[link_id].ap_chwidth == AP_AUTO)
+			dut->mld[link_id].ap_chwidth = AP_80;
 
 		free(str);
-	} else if (dut->ap_mode == AP_inval) {
-		if (dut->ap_channel <= 11)
-			dut->ap_mode = AP_11ng;
+	} else if (dut->mld[link_id].ap_mode == AP_inval) {
+		if (dut->mld[link_id].ap_channel <= 11)
+			dut->mld[link_id].ap_mode = AP_11ng;
 		else if (dut->program == PROGRAM_VHT)
-			dut->ap_mode = AP_11ac;
+			dut->mld[link_id].ap_mode = AP_11ac;
 		else
-			dut->ap_mode = AP_11na;
+			dut->mld[link_id].ap_mode = AP_11na;
 	}
 
 	/* Override the AP mode in case of 60 GHz */
 	if (dut->program == PROGRAM_60GHZ) {
-		dut->ap_mode = AP_11ad;
+		dut->mld[link_id].ap_mode = AP_11ad;
 		/* Workaround to force channel 2 if not specified */
-		if (!dut->ap_channel)
-			dut->ap_channel = 2;
+		if (!dut->mld[link_id].ap_channel)
+			dut->mld[link_id].ap_channel = 2;
 	}
 
-	switch (dut->ap_mode) {
+	switch (dut->mld[link_id].ap_mode) {
 	case AP_11g:
 	case AP_11b:
 	case AP_11ng:
@@ -1536,9 +1494,11 @@ static enum sigma_cmd_result cmd_ap_set_
 	case AP_11be:
 		if (dut->ap_band_6g)
 			dut->use_5g = 1;
-		else if (dut->ap_channel >= 1 && dut->ap_channel <= 14)
+		else if (dut->mld[link_id].ap_channel >= 1 &&
+			 dut->mld[link_id].ap_channel <= 14)
 			dut->use_5g = 0;
-		else if (dut->ap_channel >= 36 && dut->ap_channel <= 171)
+		else if (dut->mld[link_id].ap_channel >= 36 &&
+			 dut->mld[link_id].ap_channel <= 171)
 			dut->use_5g = 1;
 		break;
 	case AP_11ad:
@@ -1549,32 +1509,23 @@ static enum sigma_cmd_result cmd_ap_set_
 	if (dut->ap_is_dual)
 		dut->use_5g = 1;
 
-	if ((wlan_tag > 1 && dut->ap_band_1 == BAND_UNKNOWN) ||
-	    (dut->ap_is_mld && dut->ap_band_1 == BAND_UNKNOWN)) {
-		if (is_6g)
-			dut->ap_band_1 = BAND_6G;
-		else if (dut->ap_channel_1 >= 1 && dut->ap_channel_1 <= 14)
-			dut->ap_band_1 = BAND_2G;
-		else if (dut->ap_channel_1 >= 36 && dut->ap_channel_1 <= 171)
-			dut->ap_band_1 = BAND_5G;
-
-		if (dut->ap_band_1 != BAND_2G && (dut->ap_mode_1 == AP_11ac ||
-		    dut->ap_mode_1 == AP_11ax || dut->ap_mode == AP_11be) &&
-		    !chwidth_check)
-			dut->ap_chwidth_1 = AP_80;
-	} else if (dut->ap_band == BAND_UNKNOWN) {
-		if (is_6g)
-			dut->ap_band = BAND_6G;
-		else if (dut->ap_channel >= 1 && dut->ap_channel <= 14)
-			dut->ap_band = BAND_2G;
-		else if (dut->ap_channel >= 36 && dut->ap_channel <= 171)
-			dut->ap_band = BAND_5G;
-
-		if (dut->ap_band != BAND_2G && (dut->ap_mode == AP_11ac ||
-		    dut->ap_mode == AP_11ax || dut->ap_mode == AP_11be) &&
-		    dut->ap_80plus80 != 1)
-			dut->ap_chwidth = AP_80;
-	}
+	if (is_6g)
+		dut->mld[link_id].ap_band = BAND_6G;
+	else if (dut->mld[link_id].ap_channel >= 1 &&
+	         dut->mld[link_id].ap_channel <= 14 &&
+	         dut->mld[link_id].ap_band == BAND_UNKNOWN)
+		dut->mld[link_id].ap_band = BAND_2G;
+	else if (dut->mld[link_id].ap_channel >= 36 &&
+	         dut->mld[link_id].ap_channel <= 171 &&
+	         dut->mld[link_id].ap_band == BAND_UNKNOWN)
+		dut->mld[link_id].ap_band = BAND_5G;
+
+	if (dut->mld[link_id].ap_band != BAND_2G &&
+	    (dut->mld[link_id].ap_mode == AP_11ac ||
+	     dut->mld[link_id].ap_mode == AP_11ax ||
+	     dut->mld[link_id].ap_mode == AP_11be) &&
+	    dut->ap_80plus80 != 1 && dut->mld[link_id].ap_chwidth == AP_AUTO)
+		dut->mld[link_id].ap_chwidth = AP_80;
 
 	val = get_param(cmd, "WME");
 	if (val) {
@@ -1841,7 +1792,7 @@ static enum sigma_cmd_result cmd_ap_set_
 			get_he_mcs_nssmap((uint8_t *) &mcsnssmap, nss, mcs);
 			dut->he_mcsnssmap = (mcsnssmap << 16) | mcsnssmap;
 			/* To set mcsnssmap for EHT program */
-			if (dut->ap_mode == AP_11be)
+			if (dut->mld[link_id].ap_mode == AP_11be)
 				set_eht_mcsmap_nss(dut, nss, mcs);
 		} else {
 			set_vht_mcsmap_nss(dut, nss, mcs);
@@ -1883,9 +1834,7 @@ static enum sigma_cmd_result cmd_ap_set_
 			return STATUS_SENT;
 		}
 
-		if (dut->num_links <= MAX_MLD_SUPPORTED)
-			dut->ap_link_chwidth[dut->num_links-1] = chwidth;
-
+		dut->mld[link_id].ap_chwidth = chwidth;
 	}
 	/* TODO: WIDTH_SCAN */
 
@@ -1895,7 +1844,7 @@ static enum sigma_cmd_result cmd_ap_set_
 	dut->ap_tdls_prohibit_chswitch =
 		val && strcasecmp(val, "Enabled") == 0;
 	val = get_param(cmd, "HS2");
-	if (val && wlan_tag == 1)
+	if (val && wlan_tag <= 1)
 		dut->ap_hs2 = atoi(val);
 	val = get_param(cmd, "P2P_CROSS_CONNECT");
 	if (val)
@@ -2890,7 +2839,7 @@ static enum sigma_cmd_result cmd_ap_set_
 			char *punctchannel_val;
 			char *saveptr;
 			char *punct_channel = NULL;
-			int n, i = 0;
+			int n, i = 0, link_id_6g = get_link_id_by_band(dut, BAND_6G);
 
 			punct_channel = strdup(val);
 			if (!punct_channel) {
@@ -2898,7 +2847,7 @@ static enum sigma_cmd_result cmd_ap_set_
 			}
 
 			punctchannel_val = strtok_r(punct_channel, " ", &saveptr);
-			if (dut->ap_link_chwidth[0] == AP_80) {
+			if (dut->mld[link_id_6g].ap_chwidth == AP_80) {
 				int punct_80[] = {33, 37, 41, 45};
 
 				while (punctchannel_val != NULL) {
@@ -2910,7 +2859,7 @@ static enum sigma_cmd_result cmd_ap_set_
 					}
 					punctchannel_val = strtok_r(NULL, " ", &saveptr);
 				}
-			} else if (dut->ap_link_chwidth[0] == AP_160) {
+			} else if (dut->mld[link_id_6g].ap_chwidth == AP_160) {
 				int punct_160[] = {33, 37, 41, 45, 49, 53, 57, 61};
 
 				while (punctchannel_val != NULL) {
@@ -2972,7 +2921,7 @@ static enum sigma_cmd_result cmd_ap_set_
 		}
 	}
 
-	if (dut->ap_mode == AP_11be)
+	if (dut->mld[link_id].ap_mode == AP_11be)
 		dut->ap_beacon_prot = 1;
 
 	val = get_param(cmd, "MBSSID_MU");
@@ -3247,7 +3196,6 @@ static const char * get_sae_pk_key(const
 	return NULL;
 }
 
-
 static enum sigma_cmd_result cmd_ap_set_security(struct sigma_dut *dut,
 						 struct sigma_conn *conn,
 						 struct sigma_cmd *cmd)
--- a/sigma_dut.h
+++ b/sigma_dut.h
@@ -624,6 +624,7 @@ struct multi_mbss_config {
 	char psk[MAX_PSK_LEN];
 	unsigned char non_tx_bssid[6];
 	char ifname[IFNAMSIZ];
+	enum if_band ap_band;
 };
 
 struct mbss_support {
@@ -641,6 +642,73 @@ struct mbssid_set {
 	struct multi_mbss_config conf[MAX_NON_TX_VAP_COUNT + 1]; /* +1 for TX */
 };
 
+enum ap_chwidth {
+	AP_20,
+	AP_40,
+	AP_80,
+	AP_160,
+	AP_320,
+	AP_80_80,
+	AP_AUTO
+};
+
+enum ap_pmf {
+	AP_PMF_DISABLED,
+	AP_PMF_OPTIONAL,
+	AP_PMF_REQUIRED
+};
+
+enum ap_cipher {
+	AP_NO_GROUP_CIPHER_SET,
+	AP_CCMP,
+	AP_TKIP,
+	AP_WEP,
+	AP_PLAIN,
+	AP_CCMP_TKIP,
+	AP_GCMP_256,
+	AP_GCMP_128,
+	AP_CCMP_256,
+	AP_CCMP_128_GCMP_256,
+};
+
+enum ap_group_mgmt_cipher {
+	AP_NO_GROUP_MGMT_CIPHER_SET,
+	AP_BIP_GMAC_256,
+	AP_BIP_CMAC_256,
+	AP_BIP_GMAC_128,
+	AP_BIP_CMAC_128,
+};
+
+enum ap_mode {
+	AP_11a,
+	AP_11g,
+	AP_11b,
+	AP_11na,
+	AP_11ng,
+	AP_11ac,
+	AP_11ad,
+	AP_11ax,
+	AP_11be,
+	AP_inval
+};
+
+#define NUM_OF_LINKS 3
+
+struct mld_ap {
+	int ap_channel;
+	enum ap_chwidth ap_chwidth;
+	enum if_band ap_band;
+	int dtim;
+	unsigned int ap_akm_link;
+	enum ap_pmf ap_pmf_link;
+	enum ap_cipher ap_cipher_link, ap_group_cipher_link;
+	enum ap_group_mgmt_cipher ap_group_mgmt_cipher_link;
+	int ap_beacon_prot_link;
+	char ap_psk_link[100];
+	char *ap_sae_groups_link;
+	enum ap_mode ap_mode;
+};
+
 struct sigma_dut {
 	const char *main_ifname;
 	char *main_ifname_2g;
@@ -744,18 +812,7 @@ struct sigma_dut {
 	 * tag higher than 1 will use 'ap_tag_' variables.
 	 */
 	char ap_tag_ssid[MAX_WLAN_TAGS - 1][MAX_SSID_LEN];
-	enum ap_mode {
-		AP_11a,
-		AP_11g,
-		AP_11b,
-		AP_11na,
-		AP_11ng,
-		AP_11ac,
-		AP_11ad,
-		AP_11ax,
-		AP_11be,
-		AP_inval
-	} ap_mode;
+	enum ap_mode ap_mode;
 	int ap_channel;
 	int ap_tag_channel[MAX_WLAN_TAGS - 1];
 	int ap_rts;
@@ -784,15 +841,7 @@ struct sigma_dut {
 	unsigned int ap_vhtmcs_map;
 	enum value_not_set_enabled_disabled ap_ldpc;
 	enum value_not_set_enabled_disabled ap_sig_rts;
-	enum ap_chwidth {
-		AP_20,
-		AP_40,
-		AP_80,
-		AP_160,
-		AP_320,
-		AP_80_80,
-		AP_AUTO
-	} ap_chwidth;
+	enum ap_chwidth ap_chwidth;
 	enum ap_chwidth default_11na_ap_chwidth;
 	enum ap_chwidth default_11ng_ap_chwidth;
 	enum value_not_set_enabled_disabled ap_tx_stbc;
@@ -811,31 +860,9 @@ struct sigma_dut {
 	int ap_add_sha256;
 	int ap_add_sha384;
 	int ap_rsn_preauth;
-	/*Use ap_pmf for whole MLD level or legacy cases*/
-	enum ap_pmf {
-		AP_PMF_DISABLED,
-		AP_PMF_OPTIONAL,
-		AP_PMF_REQUIRED
-	} ap_pmf;
-	enum ap_cipher {
-		AP_NO_GROUP_CIPHER_SET,
-		AP_CCMP,
-		AP_TKIP,
-		AP_WEP,
-		AP_PLAIN,
-		AP_CCMP_TKIP,
-		AP_GCMP_256,
-		AP_GCMP_128,
-		AP_CCMP_256,
-		AP_CCMP_128_GCMP_256,
-	} ap_cipher, ap_group_cipher;
-	enum ap_group_mgmt_cipher {
-		AP_NO_GROUP_MGMT_CIPHER_SET,
-		AP_BIP_GMAC_256,
-		AP_BIP_CMAC_256,
-		AP_BIP_GMAC_128,
-		AP_BIP_CMAC_128,
-	} ap_group_mgmt_cipher;
+	enum ap_pmf ap_pmf;
+	enum ap_cipher ap_cipher, ap_group_cipher;
+	enum ap_group_mgmt_cipher ap_group_mgmt_cipher;
 	char *ap_sae_groups;
 	int sae_anti_clogging_threshold;
 	int sae_reflection;
@@ -991,7 +1018,6 @@ struct sigma_dut {
 	int ap_ul_availcap;
 	int ap_dl_availcap;
 	int ap_akm;
-	/*Use ap_akm_values for whole MLD level or legacy cases*/
 	unsigned int ap_akm_values;
 	int ap_pmksa;
 	int ap_pmksa_caching;
@@ -1361,7 +1387,6 @@ struct sigma_dut {
 	struct btm_ap btm[NUM_NEIGHBORS];
 	int btm_ml_link;
 #define MAX_MLD_SUPPORTED 2
-	enum ap_chwidth ap_link_chwidth[MAX_MLD_SUPPORTED];
 	/*Use ap_akm_link_values for interfcae specific cases*/
 	int ap_akm_link_values[MAX_MLD_SUPPORTED];
 
@@ -1371,6 +1396,9 @@ struct sigma_dut {
 	int multi_mbssid;
 	int no_of_mbssid_groups;
 	struct mbssid_set mbss_set[MAX_MBSSID_SET];
+	struct mld_ap mld[NUM_OF_LINKS];
+	int link_specific_security;
+	int multi_mld;
 };
 
 #ifdef NL80211_SUPPORT
